src/client/client.rs

```diff
@@ -87,33 +86,38 @@ impl Connection {
     /// If the connection is not yet open, then a new connection will be opened.
     /// If the request cannot be written, then a new connection is opened and the request is retried once more.
     fn send(&mut self, request: &Request) -> Result<Response, RequestError> {
-        self.try_write(request)?;
-        read_response(self.reader.as_mut().unwrap()).map_err(ResponseParsingError::into)
+        self.try_write(request).map_err(|err| Sending(err))?;
+
+        let response_parser = ResponseParser::new();
+        match response_parser.parse(self.reader.as_mut().unwrap())? {
+            Done(response) => Ok(response),
+            IoErr(_, err) => Err(Reading(err))
+        }
     }
 
     /// Tries to write the request to the server.
     /// If an existing connection is open, then that connection will be written to, otherwise a new connection is opened.
     /// If the existing connection cannot be written to, then a new connection is opened.
-    fn try_write(&mut self, request: &Request) -> Result<(), RequestError> {
+    fn try_write(&mut self, request: &Request) -> Result<(), Error> {
         self.ensure_connected()?;
         if let Ok(_) = write_request(self.writer.as_mut().unwrap(), request) {
             Ok(())
         } else {
             self.connect()?;
-            write_request(self.writer.as_mut().unwrap(), request).map_err(Error::into)
+            write_request(self.writer.as_mut().unwrap(), request)
         }
     }
 
     /// Connects to the server if not already connected.
-    fn ensure_connected(&mut self) -> Result<(), RequestError> {
+    fn ensure_connected(&mut self) -> Result<(), Error> {
         if let None = self.reader {
```

Comment:
Is this the same as self.reader == None

Reply:
Oh yeah it is, I really overuse "if let" statements haha.